sunpy/io/file_tools.py

```diff
@@ -77,19 +79,24 @@ def read_file(filepath, filetype=None, **kwargs):
     -----
     Other keyword arguments are passed to the reader used.
     """
-    filepath = str(pathlib.Path(filepath))
+    _filepath = filepath.name if isinstance(filepath, io.IOBase) else filepath
```

Comment:
Not all objects which implement `io.IOBase` have a `.name` attribute:

```python
>>> import io
>>> isinstance(io.BytesIO(), io.IOBase)
True
>>> io.BytesIO().name
AttributeError: '_io.BytesIO' object has no attribute 'name'
```

I don't think we should do this at all when dealing with io objects, I think it would be better to work out the type from the magic bytes at the start of the file and not any reference to the filename. (Which already happens if this filename doesn't match anything I think).